tobler/area_weighted/area_interpolate.py

```diff
@@ -9,10 +9,135 @@
+        yield geoms_to_chunk.iloc[start : start + chunk_size], geoms_full
+
+
+def _index_n_query(geoms1, geoms2):
+    # Pick largest for STRTree, query_bulk the smallest
+    if geoms1.shape[0] > geoms2.shape[0]:
+        large = geoms1
+        small = geoms2
+    else:
+        large = geoms2
+        small = geoms1
+    # Build tree + query
+    qry_polyIDs, tree_polyIDs = large.sindex.query_bulk(small, predicate="intersects")
+    # Remap IDs to global
+    large_global_ids = large.iloc[tree_polyIDs].index.values
+    small_global_ids = small.iloc[qry_polyIDs].index.values
+    # Return always global IDs for geoms1, geoms2
+    if geoms1.shape[0] > geoms2.shape[0]:
+        return np.array([large_global_ids, small_global_ids]).T
+    else:
+        return np.array([small_global_ids, large_global_ids]).T
+
+
+def _chunk_polys(id_pairs, geoms_left, geoms_right, n_jobs):
+    chunk_size = np.int64(id_pairs.shape[0] / n_jobs) + 1
+    for i in range(n_jobs):
+        start = i * chunk_size
+        chunk1 = geoms_left.values.data[id_pairs[start : start + chunk_size, 0]]
+        chunk2 = geoms_right.values.data[id_pairs[start : start + chunk_size, 1]]
+        yield chunk1, chunk2
+
+
+def _intersect_area_on_chunk(geoms1, geoms2):
+    import pygeos
+
+    intersection = pygeos.intersection(geoms1, geoms2)
+    areas = pygeos.measurement.area(intersection)
+    return areas
+
+
+def _area_tables_binning_parallel(source_df, target_df, n_jobs=-1):
+    """Construct area allocation and source-target correspondence tables using
+    a parallel spatial indexing approach
+    ...
+
+    NOTE: currently, the largest df is chunked and the other one is shipped in
+    full to each core; within each process, the spatial index is built for the
+    largest set of geometries, and the other one used for `query_bulk`
+
+    Parameters
+    ----------
+    source_df : geopandas.GeoDataFrame
+        GeoDataFrame containing input data and polygons
+    target_df : geopandas.GeoDataFramee
+        GeoDataFrame defining the output geometries
+    n_jobs : int
+        [Optional. Default=-1] Number of processes to run in parallel. If -1,
+        this is set to the number of CPUs available
+
+    Returns
+    -------
+    tables : scipy.sparse.dok_matrix
+
+    """
+    from joblib import Parallel, delayed, parallel_backend
+
+    if _check_crs(source_df, target_df):
+        pass
+    else:
+        return None
+    if n_jobs == -1:
+        n_jobs = os.cpu_count()
+
+    df1 = source_df.copy()
+    df2 = target_df.copy()
+
+    # Chunk the largest, ship the smallest in full
+    if df1.shape[0] > df2.shape[1]:
+        to_chunk = df1
+        df_full = df2
+    else:
+        to_chunk = df2
+        df_full = df1
+
+    # Spatial index query
+    ## Reindex on positional IDs
+    to_workers = _chunk_dfs(
+        gpd.GeoSeries(to_chunk.geometry.values, crs=to_chunk.crs),
+        gpd.GeoSeries(df_full.geometry.values, crs=df_full.crs),
```

Comment:
Why are you creating new `GeoSeries` out of existing? `gpd.GeoSeries(to_chunk.geometry.values, crs=to_chunk.crs)` should be equal to `to_chunk.geometry` (the only difference could be `name`).

Reply:
We create those `GeoSeries` to use positional indices rather than `to_chunk.index` which is not positional and may have duplicates